src/main/kotlin/com/criptext/mail/scenes/emaildetail/workers/UpdateUnreadStatusWorker.kt

```diff
@@ -22,23 +24,37 @@ class UpdateUnreadStatusWorker(
         private val db: EmailDetailLocalDB,
         private val threadId: String,
         private val updateUnreadStatus: Boolean,
+        httpClient: HttpClient,
+        activeAccount: ActiveAccount,
         private val currentLabel: Label,
         override val publishFn: (EmailDetailResult.UpdateUnreadStatus) -> Unit)
     : BackgroundWorker<EmailDetailResult.UpdateUnreadStatus> {
 
+    private val apiClient = EmailDetailAPIClient(httpClient, activeAccount.jwt)
+
     override val canBeParallelized = false
 
     override fun catchException(ex: Exception): EmailDetailResult.UpdateUnreadStatus {
         return EmailDetailResult.UpdateUnreadStatus.Failure()
     }
 
     override fun work(reporter: ProgressReporter<EmailDetailResult.UpdateUnreadStatus>): EmailDetailResult.UpdateUnreadStatus? {
-        val rejectedLabels = Label.defaultItems.rejectedLabelsByMailbox(currentLabel).map { it.id }
-        val emailIds = db.getFullEmailsFromThreadId(threadId, rejectedLabels).map {
-            it.email.id
+        val result = Result.of { apiClient.postThreadReadChangedEvent(listOf(threadId),
+                updateUnreadStatus)}
+                .mapError(HttpErrorHandlingHelper.httpExceptionsToNetworkExceptions)
+        return when (result) {
+            is Result.Success -> {
+                val rejectedLabels = Label.defaultItems.rejectedLabelsByMailbox(currentLabel).map { it.id }
+                val emailIds = db.getFullEmailsFromThreadId(threadId, rejectedLabels).map {
+                    it.email.id
+                }
+                db.updateUnreadStatus(emailIds, updateUnreadStatus)
```

Comment:
Do all database work before checking the result type

Reply:
I know that the database transaction could fail and that's why I assume you say that, but this was the agreed upon convention for all local changes that need an http post event, all database work happens only after the http post is a success (this is happening nott only in all that kind of code, but also in all clients). I was thinking maybe a transaction, but even so, the commit of it would happen in the success result of the http, what do you suggest?

Reply:
You are misunderstanding how [`Result`](https://github.com/kittinunf/Result) works.

The return type for all `Result` functions is `Result<V, E>`. That means that it is parametrized over two types: the actual value returned by the computation. and any exception thrown by the computation. Under the hood, `Result` uses sum types, so at any given point it can either hold a `V` or a `E`, but never the two at the same time. 

This works the same way as the "Except Monad" in a purely functional language like Haskell. You can "chain" more and more computations, and if the `Result` object holds a `V` value, those computations will run, producing more values. But, if it holds an `E` value, they will be ignored. This approach to error handling is more composable than the traditional try catch blocks that Kotlin inherited from Java. it's easier to add more error prone computations with less boilerplate. So, if you chain together the API call and the Database transaction, when the API call fails, the database transaction is automatically ignored.

Reply:
Awesome,thanks for explaining, I'll do just that then.